### DISCUSSION TITLE
Opt in / out of ESM image transforms

### DISCUSSION THREAD

**Discussion by Alex-S-Davies** [withastro/roadmap#765] [Proposal]
# Summary

Importing any file with an image suffix (.jpg | .png | .svg | etc) results in astro transforming that import into an ImageMetadata object. This interferes with other plugins that the author might want to use to do their own transform.

# Background & Motivation

Below is a common solution for using svgs as components in vue. Basically you pass the name of the file of the svg as a prop, and with [vite-svg-loader](https://www.npmjs.com/package/vite-svg-loader), the svg is transformed into a component, which is then used in the dynamic `component`.

```
<template>
    <component  :is="icon" />
</template>

<script setup lang="ts">
import { defineAsyncComponent } from 'vue';

export interface SvgIconProps {
    name: string,
}

const props = defineProps<SvgIconProps>();

const icon = defineAsyncComponent(
    () => import(`path/to/images/${props.name}.svg`),
);
</script>
```

However, this fails because astro transforms the import before the other plugin can do its own transform, so the above solution does not work. So we have to use the `?url` suffix escape hatch. However, that same `?url` suffix is also the same escape hatch `vite-svg-loader` uses, which basically means you can't use that plugin with astro.

The workaround:

```
<template>
    <div
        style="display: contents"
        v-html="icon"
    />
</template>

<script setup lang="ts">
import { ref, watch } from 'vue';

export interface SvgIconProps {
    name: string,
}

const props = defineProps<SvgIconProps>();

const extractSvgFile = async (name: string): Promise<string> => {
    const file = await import(`../../assets/icons/${name}.svg?raw`);
    return file.default;
};

const icon = ref(await extractSvgFile(props.name));

watch(() => props.name, async () => {
    icon.value = await extractSvgFile(props.name);
}, { immediate: true });
</script>
```

The workaround isn't horrific and is similar to how you would achieve the same result in `.astro` components. But it does result in a wrapper element, which isn't ideal and if the component is not prerendered on the server, it needs to be wrapped in a `<Suspense>` component, which means a possible bigger JS payload as the `<Suspense>` can no longer be tree shaken.

# Goals

I feel as a framework, if you are doing some _compiler magic_ you should be able to turn that magic off when it is not needed.

I feel this should especially the case in framework UI components which should be completely isolated from any astro magic (or at least opt in).

# Solutions

Maybe move the image transformation itself into the `GetImage` and `Image` components, rather than on the ESM import. I think the import transformation should be exposed as a function form the astro framework (so you can opt in if necessary), and the above functions and components can use the transformation function internally.

Maybe have a whitelist / blacklist where you can opt a folder in / out of image transformations?

**Comment by Princesseuh** [comment]
You can pass any query param to opt-out. So you can do `?vue` or something.

I'll note that this is not "compiler magic" more than Vite returning an URL or this Vue plugin returning a Vue component - it actually works the same way.

**Reply by Alex-S-Davies** [reply]
Thank you for your quick response! Sorry if I came across a little obtuse, I didn't mean to be.

Yes sorry, I did try to acknowledge that they both were using the same feature of vite to achieve their purpose, but it all seems a bit magic to me...

I have tried doing the `?vue` and even `?component` (which is the plugins way of opting into its behaviour), but I still get the same ImageMetadata astro object.

Is this more of a bug then?

**Reply by Princesseuh** [reply]
Hmm, make sure you're on the latest version of Astro. We definitely ignore imports with query params, here's a repro: https://stackblitz.com/edit/github-cwpkze?file=src%2Fpages%2Findex.astro

Additionally, we have a test in our codebase using the plugin you cited in your original message: https://github.com/withastro/astro/blob/5f2db426ce13d8b3ad6f506490aec8a980e48cc7/packages/integrations/vue/test/fixtures/app-entrypoint/src/components/Foo.vue and it works

**Reply by Alex-S-Davies** [reply]
I have just updated Astro and can confirm the issue is still present.

Using your repro, and editing it slightly results is what I am experiencing:
https://stackblitz.com/edit/github-cwpkze-aajfwc?file=src%2Fpages%2Findex.astro

So it looks like something to do with dynamic imports? Should I open an issue?

Regardless, this should be closed, as clearly there are opt outs already.

**Reply by Princesseuh** [reply]
Ah yeah, this is a bug with dynamic imports in Vite 4.x. https://github.com/vitejs/vite/issues/14449

We'll be upgrading to Vite 5 for Astro 4 (coming soon), but until then, I'm not sure if there's a way to work around this.

**Reply by Alex-S-Davies** [reply]
Ahh, that is no problem!

I do have a workaround that I can use for now, but I will update as soon as I can.
Thanks for looking into this for me, you guys do great work!
